core/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -84,6 +85,7 @@ public GatewayMetaState(Settings settings, NodeEnvironment nodeEnv, MetaStateSer
         if (DiscoveryNode.masterNode(settings) || DiscoveryNode.dataNode(settings)) {
             try {
                 ensureNoPre019State();
+                IndexFolderUpgrader.upgradeIndicesIfNeeded(settings, nodeEnv);
                 pre20Upgrade();
```

Comment:
not sure about this - the pre20upgrade thing checks for old versions of state files and issues the proper message (via MetaDataIndexUpgradeService#checkSupportedVersion ). I think we should check for too old versions before starting to mess up with folders. People trying to go from a 1.x cluster to 5.0 should be rejected before we start messing up with folders.